### PR TITLE
[PixArt-Alpha] Introduce resolution binning

### PR DESCRIPTION
**Pull request by sayakpaul**
# What does this PR do?

Introduces resolution binning as originally done by @lawrence-cj in https://github.com/huggingface/diffusers/pull/5716. 

This is particularly useful for generating non-square images. 

### 1024x2048

| No Binning | Binning |
|:---:|:---:|
|  |  |
|  |  |
|  |  |
|  |  |

### 1536x720

| No Binning | Binning |
|:---:|:---:|
|  |  |
|  |  |
|  |  |
|  |  |

## Code

```python
import torch

from diffusers import PixArtAlphaPipeline

prompts = [
    "A small cactus with a happy face in the Sahara desert.",
    "Pirate ship trapped in a cosmic maelstrom nebula, rendered in cosmic beach whirlpool engine, volumetric lighting, spectacular, ambient lights, light pollution, cinematic atmosphere, art nouveau style, illustration art artwork by SenseiJaye, intricate detail.",
    "stars, water, brilliantly, gorgeous large scale scene, a little girl, in the style of dreamy realism, light gold and amber, blue and pink, brilliantly illuminated in the background.",
    "nature vs human nature, surreal, UHD, 8k, hyper details, rich colors, photograph.",
]

pipe = PixArtAlphaPipeline.from_pretrained(
    "PixArt-alpha/PixArt-XL-2-1024-MS", torch_dtype=torch.float16
)
pipe.enable_model_cpu_offload()

no_resolution_binning = []
resolution_binning = []

for i, prompt in enumerate(prompts):
    generator = torch.manual_seed(2344) 
    image = pipe(prompt, height=1536, width=720, generator=generator, use_resolution_binning=False).images[0]
    image.save(f"no_resolution_binning_{i}_1536x720.png")
    no_resolution_binning.append(image)
    
    generator = torch.manual_seed(2344)
    image = pipe(prompt, height=1536, width=720, generator=generator).images[0]
    image.save(f"resolution_binning_{i}_1536x720.png")
    resolution_binning.append(image)

```

### CODE REVIEW

**Comment by sayakpaul** [comment]
@DN6 we're introducing a new dependency here - `torchvision`. Could you please help reviewing if I added all the necessary guardrails to not break anything?

**Comment by lawrence-cj** [comment]
No problem on my side.

**Comment by adhikjoshi** [comment]
can we enable "resolution_binning" by default?

looks like a good approach to have it enabled as default,

File: src/diffusers/pipelines/pixart_alpha/pipeline_pixart_alpha.py [hunk 1]
```diff
@@ -495,6 +534,28 @@ def prepare_latents(self, batch_size, num_channels_latents, height, width, dtype
...
+    def resize_and_crop_tensor(samples: torch.Tensor, new_width: int, new_height: int) -> torch.Tensor:
+        orig_hw = torch.tensor([samples.shape[2], samples.shape[3]])
+        custom_hw = torch.tensor([new_height, new_width])
+
+        if (orig_hw != custom_hw).all():
+            ratio = max(custom_hw[0] / orig_hw[0], custom_hw[1] / orig_hw[1])
+            resized_width = int(orig_hw[1] * ratio)
+            resized_height = int(orig_hw[0] * ratio)
+            transform = T.Compose([T.Resize((resized_height, resized_width)), T.CenterCrop(custom_hw.tolist())])
+            return transform(samples)
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/pixart_alpha/pipeline_pixart_alpha.py:555]: This is not worth a `torchvision` dependency IMO. `torchvision` is very heavy and we can do this operations without torchvision

File: src/diffusers/pipelines/pixart_alpha/pipeline_pixart_alpha.py [hunk 2]
```diff
@@ -43,6 +44,8 @@
 if is_ftfy_available():
     import ftfy
 
+if is_torchvision_available():
+    from torchvision import transforms as T
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/pixart_alpha/pipeline_pixart_alpha.py:48]:
```suggestion
```

**Review by patrickvonplaten — COMMENTED** [review]
Very nice addition! Let's add a test as well and please no `torchvision` dependency :pray:

**Comment by sayakpaul** [comment]
Eliminated torchvision as a dependency and the results look fine:

### 1024x2048

| No Binning | Binning |
|:---:|:---:|
|  |  |
|  |  |
|  |  |
|  |  |

### 1536x720

| No Binning | Binning |
|:---:|:---:|
|  |  |
|  |  |
|  |  |
|  |  |

**Comment by sayakpaul** [comment]
@patrickvonplaten the reason why we cannot do a fast test for resolution binning is that even for very resolutions, `classify_height_width_bin()` would return extremely high resolutions to run on a CPU. 

This is why I choose to add a slow test for it and include that in https://github.com/huggingface/diffusers/pull/5752 (as it requires changing the assertion values anyway). Is that okay?

**Review by patrickvonplaten — APPROVED** [review]
Clean!

**Pull request merged**